### PR TITLE
Add Meet Data

### PR DESCRIPTION
**Pull request by denglender**
Wrote the code to add meet data and tested on my domain.  Working fine for me.

### CODE REVIEW

**Comment by denglender** [comment]
resolves #65

**Comment by dchess** [comment]
@denglender Thanks for contributing! I'll give this a go with our data in testing. @zkagin Can you take a look at adding tests for this as well?

File: README.md [hunk 1]
```diff
@@ -114,6 +116,7 @@ https://www.googleapis.com/auth/classroom.profile.emails,
 https://www.googleapis.com/auth/classroom.rosters,
 https://www.googleapis.com/auth/classroom.student-submissions.students.readonly,
 https://www.googleapis.com/auth/classroom.topics
```
Comment (dchess) [inline README.md:118]: @denglender This needs a comma at the end in order to properly copy & paste

**Comment by dchess** [comment]
@denglender This looks good overall and ran in our environment in less than 7 mins. However, it's currently not passing our checks. Can you try formatting your code with [black](https://pypi.org/project/black/)?

File: api.py [hunk 2]
```diff
@@ -701,3 +701,53 @@ def request_data(self, course_id=None, date=None, next_page_token=None):
...
+
+    def preprocess_records(self, records):
+        """Parse the coursework nested json into flat records for insertion
+        in to database table"""
+        new_records = []
+        for record in records:
+            eventrecords = record.get("events")
+            curtime = record.get("id").get("time")
+            for eventrecord in eventrecords:
+                if eventrecord.get("name") == "call_ended":
```
Comment (zkagin) [inline api.py:742]: If the params only pull this event, do we need this?

File: api.py [hunk 3]
```diff
@@ -701,3 +701,49 @@ def request_data(self, course_id=None, date=None, next_page_token=None):
...
+        new_records = []
+        for record in records:
+            eventrecords = record.get("events")
+            curtime = record.get("id").get("time")
+            for eventrecord in eventrecords:
+                if eventrecord.get("name") == "call_ended":                    
+                    cur = {"item_time":curtime}
+                    for subrecord in eventrecord.get("parameters"):
+                        if subrecord.get("name") in self.columns:
+                            cur[subrecord.get("name")] = subrecord.get("value",subrecord.get("intValue",subrecord.get("boolValue")))
```
Comment (zkagin) [inline api.py]: This is a pretty hard line to parse, I think because of the nesting (though the word wrap. Perhaps something like below would be easier to read?
name = subrecord.get("name")
value = subrecord.get("value") or subrecord.get("intValue") or subrecord.get("boolValue")
cur[name] = value

File: api.py [hunk 4]
```diff
@@ -701,3 +701,51 @@ def request_data(self, course_id=None, date=None, next_page_token=None):
...
+            curtime = record.get("id").get("time")
+            for eventrecord in eventrecords:
+                if eventrecord.get("name") == "call_ended":
+                    cur = {"item_time": curtime}
+                    for subrecord in eventrecord.get("parameters"):
+                        if subrecord.get("name") in self.columns:
+                            cur[subrecord.get("name")] = subrecord.get(
+                                "value",
+                                subrecord.get("intValue", subrecord.get("boolValue")),
+                            )
```
Comment (zkagin) [inline api.py:749]: This line is pretty hard to parse, I think because of the nesting. What are your thoughts on something like this?
name = subrecord.get("name")
value = subrecord.get("value") or subrecord.get("intValue") or subrecord.get("boolValue")
cur[name] = value

File: api.py [hunk 5]
```diff
@@ -701,3 +701,51 @@ def request_data(self, course_id=None, date=None, next_page_token=None):
...
+        return self.service.activities().list(**options)
+
+    def preprocess_records(self, records):
+        new_records = []
+        for record in records:
+            eventrecords = record.get("events")
+            curtime = record.get("id").get("time")
+            for eventrecord in eventrecords:
+                if eventrecord.get("name") == "call_ended":
+                    cur = {"item_time": curtime}
```
Comment (zkagin) [inline api.py:743]: Can we change the variable name to something a little more explanatory — perhaps `new_item` or `new_record`?

File: api.py [hunk 6]
```diff
@@ -701,3 +701,51 @@ def request_data(self, course_id=None, date=None, next_page_token=None):
...
+            "userKey": "all",
+            "eventName": "call_ended",
+            "pageToken": next_page_token,
+        }
+        return self.service.activities().list(**options)
+
+    def preprocess_records(self, records):
+        new_records = []
+        for record in records:
+            eventrecords = record.get("events")
```
Comment (zkagin) [inline api.py:739]: Can we change this to `event_records`? I think PEP8 recommends underscores between words for variables.

File: api.py [hunk 7]
```diff
@@ -701,3 +701,51 @@ def request_data(self, course_id=None, date=None, next_page_token=None):
...
+            curtime = record.get("id").get("time")
```
Comment (zkagin) [inline api.py:740]: Can we change this to something like `current_time` or `item_time`?

File: api.py [hunk 8]
```diff
@@ -701,3 +701,51 @@ def request_data(self, course_id=None, date=None, next_page_token=None):
...
+            for eventrecord in eventrecords:
```
Comment (zkagin) [inline api.py:741]: Can we change this to `event_record`?

File: api.py [hunk 9]
```diff
@@ -701,3 +701,53 @@ def request_data(self, course_id=None, date=None, next_page_token=None):
...
+        in to database table"""
+        new_records = []
+        for record in records:
+            eventrecords = record.get("events")
+            curtime = record.get("id").get("time")
+            for eventrecord in eventrecords:
+                if eventrecord.get("name") == "call_ended":
+                    cur = {"item_time": curtime}
+                    for subrecord in eventrecord.get("parameters"):
+                        if subrecord.get("name") in self.columns:
```
[the remark this answers is not in the document]
Comment (denglender) [reply api.py:745]: I didn't look extensively at how the other data sets are returned, but the meet data is returned in a strange format that it didn't look like the existing code handled.  But let me know if I'm mistaken.  Here's an abridged example of the returned JSON.

```
{
 "kind": "admin#reports#activities",
 "items": [
  {
   "kind": "admin#reports#activity",
   "id": {
    "time": "2020-05-08T18:32:03.129Z",
    "applicationName": "meet",
   },
   "actor": {
    "callerType": "USER",
    "email": "denglender@dcprep.org",
   },
   "events": [
    {
     "type": "call",
     "name": "call_ended",
     "parameters": [
      {
       "name": "video_send_seconds",
       "intValue": "1913"
      },
      {
       "name": "audio_send_bitrate_kbps_mean",
       "intValue": "1"
      },
      [...]
```

File: api.py (hunk 2 above)
Comment (denglender) [reply api.py:742]: I included as a reminder to me that if I wanted to add in livestream data in the future I'd need to rework the code to add a column for event type.  But I can remove if you prefer.

File: api.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (zkagin) [reply api.py:745]: Yep, it is a little different than the others. To clarify on requested change, I am referencing the line:
```
if subrecord.get("name") in self.columns:
```
In other endpoints generally all of the data is pulled in, and then once it is in dataframe it is filtered to only include the columns specified in `self.columns`:
```
api.py:67
df = df.reindex(columns=self.columns)
```
So I think we can remove the if statement here, as it is handled elsewhere already?

File: api.py (hunk 2 above)
Comment (zkagin) [reply api.py:742]: Got it, makes sense. Perhaps we should add the event_type column now in that case? Even though it is constant now, it'll future-proof it for later and avoid a schema migration if they are building something on top of the data coming in.

File: api.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (denglender) [reply api.py:745]: Got it.  Didn't realize they were scrubbed elsewhere.  I'm happy to remove the line, though it may create measureable RAM usage to hold onto the 50+ (and perhaps more in the future?) data points for the 1000 rows of data.

File: api.py (hunk 2 above)
Comment (denglender) [reply api.py:742]: Added

File: api.py (hunk 4 above)
Comment (denglender) [reply api.py:749]: Sure.  black reformatted it further in a way that I find annoying, but so be it.

File: api.py (hunk 6 above)
Comment (denglender) [reply api.py:739]: Sure

File: api.py (hunk 7 above)
Comment (denglender) [reply api.py:740]: Done

File: api.py (hunk 8 above)
Comment (denglender) [reply api.py:741]: Done

File: api.py (hunk 5 above)
Comment (denglender) [reply api.py:743]: Done

File: api.py [hunk 10]
```diff
@@ -701,3 +701,53 @@ def request_data(self, course_id=None, date=None, next_page_token=None):
...
+            "item_time",
+        ]
+        self.request_key = "items"
+        self.batch_size = config.MEET_BATCH_SIZE
+
+    def request_data(self, course_id=None, date=None, next_page_token=None):
+        options = {
+            "applicationName": "meet",
+            "userKey": "all",
+            "eventName": "call_ended",
```
[the remark this answers is not in the document]
Comment (denglender) [reply api.py:731]: Added a few docstrings.  Defer to you all if you prefer named in some other way.

**Comment by awertman** [comment]
@denglender out of curiosity, did you notice any ability to connect meetings to specific courses? If not, I suspect they will release some type of meet id in the course API.

https://support.google.com/edu/classroom/answer/6149237?hl=en

**Comment by denglender** [comment]
@awertman - I didn't look, but my guess is that it's not possible currently.  I can't speak to whether the Classroom API allows you to pull the Meet link, but the Meet links in Classroom use "nicknames", and I think only the meet code is recorded in the audit data.  A nickname will map to different meet codes over time.  The way I'm relating the audit data to "attendance" is by considering a meet between a student(s) and a teacher (determined based on their usernames) as being a meeting to be counted for "attendance".

File: api.py [hunk 11]
```diff
@@ -701,3 +701,57 @@ def request_data(self, course_id=None, date=None, next_page_token=None):
...
+
+    def preprocess_records(self, records):
+        """Pull out parameter data from the returned Google Meet call event"""
+        new_records = []
+        for record in records:
+            event_records = record.get("events")
+            item_time = record.get("id").get("time")
+            for event_record in event_records:
+                event_name = event_record.get("name")
+                if event_name == "call_ended":
```
[the remark this answers is not in the document]
Comment (dchess) [reply api.py:746]: @zkagin Since this is a non-blocking issue, why don't we merge it for now? We can confirm with our own data and open a refactor branch later.
